### PR TITLE
ledger: refactor store module interfaces before kv impl merge

### PR DESCRIPTION
**Pull request by icorderi**
## Summary

We are breaking down #5212 in several smaller PRs to make it more digestible. 

This is PRs "numero 2" on the KV epic.

Content:
- Refactor of the storage interface (see [ledger/store/trackerdb/store.go](https://github.com/algorand/go-algorand/pull/5451/files#diff-0ba7f16de50be72300e516ee5c401f677bda6453b30817d7804edd97cac84f42))
- Any SQL code changes that were part of the KV work
- All ledger changes accumulated on the KV work (mostly due to the interface changes) 

This is built on top of the following PRs:
- configs: #5369 

## Test Plan

existing tests.

### CODE REVIEW

File: ledger/store/trackerdb/sqlitedriver/sqlitedriver.go [hunk 1]
```diff
@@ -0,0 +1,339 @@
...
+}
+
+func (s *trackerSQLStore) SnapshotContext(ctx context.Context, fn trackerdb.SnapshotFn) (err error) {
+	return s.pair.Rdb.AtomicContext(ctx, func(ctx context.Context, tx *sql.Tx) error {
+		return fn(ctx, sqlSnapshotScope{tx, &sqlReader{tx}})
+	})
+}
+
+func (s *trackerSQLStore) BeginSnapshot(ctx context.Context) (trackerdb.Snapshot, error) {
+	handle, err := s.pair.Wdb.Handle.BeginTx(ctx, nil)
```
Comment (algorandskiy) [inline ledger/store/trackerdb/sqlitedriver/sqlitedriver.go:93]:
```suggestion
	handle, err := s.pair.Rdb.Handle.BeginTx(ctx, nil)
```

**Review by algorandskiy — COMMENTED** [review]
Looks good, few comments

File: ledger/catchpointtracker.go [hunk 2]
```diff
@@ -1316,7 +1316,7 @@ func (ct *catchpointTracker) GetCatchpointStream(round basics.Round) (ReadCloseS
 	ledgerGetcatchpointCount.Inc(nil)
 	// TODO: we need to generalize this, check @cce PoC PR, he has something
 	//       somewhat broken for some KVs..
-	err := ct.dbs.Snapshot(func(ctx context.Context, tx trackerdb.SnapshotScope) (err error) {
+	err := ct.dbs.Transaction(func(ctx context.Context, tx trackerdb.TransactionScope) (err error) {
```
[the remark this answers is not in the document]
Comment (icorderi) [reply ledger/catchpointtracker.go:1319]: I moved the catchpoint things to be only available on a full writer for now. 
This was in preparation for the refactor.

[the remark this answers is not in the document]
Comment (algorandskiy) [reply ledger/catchpointtracker.go:1319]: that maybe MakeCatchpointReader should be called MakeCatchpointReaderWriter?

**Review by algorandskiy — APPROVED** [review]

File: ledger/acctdeltas.go [hunk 3]
```diff
@@ -344,6 +344,10 @@ func (a *compactResourcesDeltas) resourcesLoadOld(tx trackerdb.TransactionScope,
 				err = fmt.Errorf("empty resource record: addrid=%d, aidx=%d", acctRef, aidx)
 				return err
 			}
+		case trackerdb.ErrNotFound:
+			// we don't have that account, just return an empty record.
+			a.updateOld(missIdx, trackerdb.PersistedResourcesData{AcctRef: acctRef, Aidx: aidx})
+			err = nil
```
Comment (AlgoAxel) [inline ledger/acctdeltas.go:350]: since theese appear identical, you could make this a fallthrough

File: ledger/acctdeltas_test.go [hunk 4]
```diff
@@ -2233,20 +2191,20 @@ func TestAccountOnlineQueries(t *testing.T) {
 
 	proto := config.Consensus[protocol.ConsensusCurrentVersion]
 
-	dbs, _ := sqlitedriver.DbOpenTrackerTest(t, true)
-	dbs.SetLogger(logging.TestingLog(t))
```
Comment (AlgoAxel) [inline ledger/acctdeltas_test.go:2237]: note to self: SetLogger being removed as it's now part of construction

File: ledger/ledger.go [hunk 5]
```diff
@@ -302,20 +296,32 @@ func openLedgerDB(dbPathPrefix string, dbMem bool) (trackerDBs trackerdb.Tracker
...
-		trackerDBs, lerr = sqlitedriver.OpenTrackerSQLStore(trackerDBFilename, dbMem)
+		switch cfg.StorageEngine {
+		case "sqlite":
+			fallthrough
+		// anything else will initialize a sqlite engine.
+		default:
+			file := dbPathPrefix + ".tracker.sqlite"
+			trackerDBs, lerr = sqlitedriver.Open(file, dbMem, log)
+		}
+
```
Comment (AlgoAxel) [inline ledger/ledger.go:310]: makes sense we wouldn't have `pebbledb` here *yet*, but the comment is a little confusing, "anything else will initialize a sqlite engine"

**Review by AlgoAxel — APPROVED** [review]
Largely, the changes here rename structs and instances. It also reorganizes the sql specific code into the sqlite driver.

I left some comments which are non blocking suggestions. I am using this code already as part of my own work (via kv-base branch) so I am comfortable with the changes as well.

**Pull request merged**